cip-gege251-reliably-ordered-tx-inputs.md

```diff
@@ -0,0 +1,73 @@
+However if the transaction inputs would be stored as an array, preserving the original ordering, we could better optimise our scripts by determining the order of inputs off-chain and simplifying the search to a pattern match.
+For example, for a script needing transaction inputs for two purposes: one for updating the datum of the script, and one or more for supplying the funds. We could order these inputs such that the one including the datum comes first, and then we can use a pattern match `txWithDatum : txWithFunds = txInfoInputs`.
+
+## Specification
+
+In the CDDL definiotion of the transaction body, transaction inputs will need to be changed from Set to Array:
+
+```
+transaction_body =
+ { 0 : [* transaction_input]     ; inputs
+ , 1 : [* transaction_output]
+ , 2 : coin                      ; fee
+ , ? 3 : uint                    ; time to live
+ , ? 4 : [* certificate]
+ , ? 5 : withdrawals
+ , ? 6 : update
+ , ? 7 : auxiliary_data_hash
+ , ? 8 : uint                    ; validity interval start
+ , ? 9 : mint
+ , ? 11 : script_data_hash       ; New
+ , ? 13 : set<transaction_input> ; Collateral ; new
+ , ? 14 : required_signers       ; New
+ , ? 15 : network_id             ; New
+ }
+```
+
+Note that the Set is defined in the cardano-ledger specification as an Array of zero or more elements `set = [* a]`, so this would only cause a semantic change, without the actual change of the underlying data structure.
+
+## Rationale
+
+In theory we could pre-calculate the index of the searched transaction input off-chain, by relying on the ordering of transaction inputs (lexicographic ordering on `(TransactionID, TransactionIndex)`). But then, to pass this information to the script, we would need to use a redeemer, which could defeat the purpose of this optimisation with it's own fees. Also, we could only reliably do this after transaction balancing, but changing the transaction after balancing could affect fees, rendering the transaction unbalanced again. This method also introduces unnecessary complexity.
+
+## Security considerations
+
+We considered the possibility of a man-in-the-middle attack, where the adversary could change the transaction input ordering, and altering the behaviour of a script. However the existing ledger rules for deriving txid are sufficient to guarantee the integrity of transactions, because the inputs (whether a set or a list) are included in the transaction body that gets hashed to a txid. Therefore, no changes would be required to txid derivation.
+
+## Backward Compatibility
+
+For ledger transactions before this CIP is implemented, the underlying CBOR data structure only changes semantically, and the difference between input order at submission and at script evaluation is irrelevant, because the scripts in those transactions do not rely on pattern matching by position. Whereas, transactions after this CIP could optionally do positional pattern matching.
```

Comment:
> because the scripts in those transactions do not rely on pattern matching by position

This is false. They perfectly well can do, it's weird, but people write lots of weird stuff.

This CIP changes the semantics of the transaction context given to scripts in an observable fashion, and as such it need a new Plutus language version, I'm afraid (see https://github.com/cardano-foundation/CIPs/pull/215).

Reply:
When is the next Plutus language version planned to be released? Will the June hardfork be Plutus V1 or Plutus V2?

Reply:
June HF will be V2, but it is too late for additional changes to go into that.

Reply:
So, this CIP would have to wait for V3?

Reply:
Yes.